odxtools/templates/macros/printComparam.xml.jinja2

```diff
@@ -0,0 +1,75 @@
+{#- -*- mode: sgml; tab-width: 1; indent-tabs-mode: nil -*-
+ #
+ # SPDX-License-Identifier: MIT
+ # Copyright (c) 2022 MBition GmbH
+-#}
+
+{# print either a simple or a complex comparam
+ #}
+{%- macro printGenericComparam(cp) %}
+{#
+ # This is slightly hacky: To distinguish complex and simple
+ # communication parameters, we check if the 'cp' object has a
+ # 'comparams' attribute. If it does, it is a complex parameter.
+ #}
+{%- if hasattr(cp, 'comparams') %}
+{{- printComplexComparam(cp) }}
+{%- else %}
+{{- printSimpleComparam(cp) }}
+{%- endif %}
+{%- endmacro %}
+
+
+{%- macro printSimpleComparam(cp) %}
+<COMPARAM ID="{{cp.odx_id.local_id}}"
+          PARAM-CLASS="{{cp.param_class}}"
+          CPTYPE="{{cp.cptype}}"
+          {%- if cp.display_level is not none %}
+          DISPLAY-LEVEL="{{cp.display_level}}"
+          {%- endif %}
+          CPUSAGE="{{cp.cpusage}}">
+ <SHORT-NAME>{{cp.short_name}}</SHORT-NAME>
+ {%- if cp.long_name and cp.long_name.strip() %}
+ <LONG-NAME>{{cp.long_name|e}}</LONG-NAME>
+ {%- endif %}
+ {%- if cp.description and cp.description.strip() %}
+ <DESC>
+ {{cp.description}}
+ </DESC>
+ {%- endif %}
+ <PHYSICAL-DEFAULT-VALUE>{{cp.physical_default_value}}</PHYSICAL-DEFAULT-VALUE>
+ <DATA-OBJECT-PROP-REF ID-REF="{{cp.dop_ref.ref_id}}" />
+</COMPARAM>
+{%- endmacro %}
+
+{%- macro printComplexComparam(cp) %}
+<COMPLEX-COMPARAM ID="{{cp.odx_id.local_id}}"
+                  PARAM-CLASS="{{cp.param_class}}"
+                  CPTYPE="{{cp.cptype}}"
+                  {%- if cp.display_level is not none %}
+                  DISPLAY-LEVEL="{{cp.display_level}}"
+                  {%- endif %}
+                  CPUSAGE="{{cp.cpusage}}"
+                  {%- if cp.allow_multiple_values == True %}
+                  ALLOW-MULTIPLE-VALUES="true"
+                  {%- elif cp.allow_multiple_values == False %}
+                  ALLOW-MULTIPLE-VALUES="false"
+                  {%- endif %}
+                  >
+ <SHORT-NAME>{{cp.short_name}}</SHORT-NAME>
+ {%- if cp.long_name and cp.long_name.strip() %}
+ <LONG-NAME>{{cp.long_name|e}}</LONG-NAME>
+ {%- endif %}
+ {%- if cp.description and cp.description.strip() %}
+ <DESC>
+ {{cp.description}}
+ </DESC>
+ {%- endif %}
+ {%- for sub_cp in cp.comparams %}
+ {{- printGenericComparam(sub_cp) | indent(1, first=True) }}
+ {%- endfor %}
+ {%- if cp.physical_default_value is not none %}
+ <COMPLEX-PHYSICAL-DEFAULT-VALUE>{{cp.physical_default_value}}</COMPLEX-PHYSICAL-DEFAULT-VALUE>
```

Comment:
not valid

Reply:
hm, I see: this is one of these overcomplicated constructs again. what got me confused is that ComparamSubset `physical_default_value()` folds this into `physical_default_value` instead of `complex_physical_default_value`. I think these two should be separated because the first is for simple parameters while the second is for complex ones?

Reply:
Using complex_physical_default_value seemed like an overkill, since physical_default_value can support multiple types (based on the dop physical type) and it would force users to do extra type checking for a common property

Reply:
If I got this right, the typing of this looks something like this:

```
SimpleValue = str
ComplexValue = List[Union[SimpleValue, ComplexValue]]
```

IMO, conflating these into the same attribute would make things quite confusing (read: I was quite confused), and also, the `COMPLEX-COMPARAM` and `COMPARAM` tags also use `COMPLEX-PHYSICAL-DEFAULT-VALUE` vs `PHYSICAL-DEFAULT-VALUE`. I've adapted the parameter extraction code in the [latest commit](https://github.com/mercedes-benz/odxtools/pull/96/commits/c71e1d19cbff01957b5bd0615172ad3e3da3a1c9), and I think the new version is easier to read. (and also takes the default values into account...)